pkg/provisioner/crdprovisioner.go

```diff
@@ -737,11 +669,54 @@ func (c *CrdProvisioner) waitForLun(ctx context.Context, volumeID, nodeID string
 	return azVolumeAttachment.Status.Detail.PublishContext, err
 }
 
+// WaitForAttachComplete waits for either AzVolumeAttachment to be complete attach or be deleted.
+func (c *CrdProvisioner) WaitForAttachComplete(ctx context.Context, volumeID, nodeID string) (bool, error) {
+	var err error
+	ctx, w := workflow.New(ctx)
+	defer func() { w.Finish(err) }()
+
+	var volumeName string
+	volumeName, err = azureutils.GetDiskName(volumeID)
+	if err != nil {
+		return false, err
+	}
+	attachmentName := azureutils.GetAzVolumeAttachmentName(volumeName, nodeID)
+
+	waiter := c.conditionWatcher.NewConditionWaiter(ctx, watcher.AzVolumeAttachmentType, attachmentName, waitForAttachVolumeCompleteFunc)
+	defer waiter.Close()
+
+	_, err = waiter.Wait(ctx)
+	if err != nil {
+		return false, err
+	}
+	return true, nil
+}
+
 func (c *CrdProvisioner) WaitForAttach(ctx context.Context, volumeID, nodeID string) (*azdiskv1beta2.AzVolumeAttachment, error) {
 	var err error
 	ctx, w := workflow.New(ctx, workflow.WithCaller(1))
 	defer func() { w.Finish(err) }()
 
+	volumeName, err := azureutils.GetDiskName(volumeID)
+	if err != nil {
+		return nil, err
+	}
+	volumeName = strings.ToLower(volumeName)
+
+	var replicasToDetach []*azdiskv1beta2.AzVolumeAttachment
+	if replicasToDetach, err = c.getReplicasToDetach(ctx, volumeName, nodeID); err != nil {
```

Comment:
Hmm... Do we want to do all this work every time we wait for lun or attachment to complete?